File: mns-reactor/src/test/java/cz/fb/manaus/reactor/betting/listener/ProbabilityComparatorTest.java
package cz.fb.manaus.reactor.betting.listener;

import cz.fb.manaus.core.model.RunnerPrices;
import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.core.test.AbstractLocalTestCase;
import cz.fb.manaus.reactor.ReactorTestFactory;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.List;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class ProbabilityComparatorTest extends AbstractLocalTestCase {

    @Autowired
    private ReactorTestFactory testFactory;

    @Test
    public void testCompare() {
        var first = testFactory.newRP(1, 1.4d, 1.6d);
        var second = testFactory.newRP(2, 2.8d, 3.3d);
        assertThat(getFirstSelection(List.of(first, first)), is(1L));
        assertThat(getFirstSelection(List.of(first, second)), is(1L));
        assertThat(getFirstSelection(List.of(second, first)), is(1L));
        assertThat(getFirstSelection(List.of(second, second)), is(2L));
    }

    private long getFirstSelection(List<RunnerPrices> lists) {
        return ProbabilityComparator.COMPARATORS.get(Side.BACK).immutableSortedCopy(lists).get(0).getSelectionId();
    }
}

File: mns-reactor/src/test/java/cz/fb/manaus/reactor/betting/listener/ProbabilityComparatorTest.kt
package cz.fb.manaus.reactor.betting.listener

import cz.fb.manaus.core.model.RunnerPrices
import cz.fb.manaus.core.model.Side
import cz.fb.manaus.reactor.PricesTestFactory
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

class ProbabilityComparatorTest {

    @Test
    fun `probability comparison`() {
        val first = PricesTestFactory.newRunnerPrices(1, 1.4, 1.6)
        val second = PricesTestFactory.newRunnerPrices(2, 2.8, 3.3)
        assertEquals(1L, getFirstSelection(listOf(first, first)))
        assertEquals(1L, getFirstSelection(listOf(first, second)))
        assertEquals(1L, getFirstSelection(listOf(second, first)))
        assertEquals(2L, getFirstSelection(listOf(second, second)))
    }

    private fun getFirstSelection(lists: List<RunnerPrices>): Long {
        return lists.sortedWith(ProbabilityComparator.COMPARATORS.getValue(Side.BACK))[0].selectionId
    }
}